sql/executor.go

```diff
@@ -528,39 +529,77 @@ func (gp golangParameters) Arg(name string) (parser.Datum, bool) {
 	if arg == nil {
 		return parser.DNull, true
 	}
-	// TODO: This type switch has been created primarily for convenience, but
-	// the fact that multiple integer types are part of the switch is inelegant
-	// and prone to errors (it has been described as "whack-a-mole"). There is
-	// likely a better way to capture all of the various integer types using
-	// reflection.
+
+	// A type switch to quickly handle common primitive types.
 	switch t := arg.(type) {
+	// Datums are passed along as is.
+	case parser.Datum:
+		return t, true
+
+	// Time datatypes get special representation in the database.
+	case time.Time:
+		return parser.DTimestamp{Time: t}, true
+	case time.Duration:
+		return parser.DInterval{Duration: t}, true
+
 	case bool:
 		return parser.DBool(t), true
+
+	// Primitive integer types.
 	case int:
 		return parser.DInt(t), true
-	case int32:
+	case int8:
 		return parser.DInt(t), true
-	case uint32:
+	case int16:
+		return parser.DInt(t), true
+	case int32:
 		return parser.DInt(t), true
 	case int64:
 		return parser.DInt(t), true
-	case uint64:
+	case uint:
 		return parser.DInt(t), true
-	case ID:
+	case uint8:
 		return parser.DInt(t), true
-	case []byte:
-		return parser.DBytes(t), true
+	case uint16:
+		return parser.DInt(t), true
+	case uint32:
+		return parser.DInt(t), true
+	case uint64:
+		return parser.DInt(t), true
+
+	// Primitive float types.
+	case float32:
+		return parser.DFloat(t), true
+	case float64:
+		return parser.DFloat(t), true
+
 	case string:
 		return parser.DString(t), true
-	case time.Time:
-		return parser.DTimestamp{Time: t}, true
-	case time.Duration:
-		return parser.DInterval{Duration: t}, true
-	case parser.Datum:
-		return t, true
-	default:
-		panic(fmt.Sprintf("unexpected type %T", t))
+	case []byte:
+		return parser.DBytes(t), true
 	}
+
+	// Handle non-primitive types which have a supported underlying type.
+	val := reflect.ValueOf(arg)
```

Comment:
Yes, this would handle primitive types as well. Performance is a concern everywhere, though we probably shouldn't be optimizing until we have a benchmark that shows this code to be problematic. I'd be fine with the type-switch above only handling `Datum`, `time.Time` and `time.Duration` with a comment that additional types could be added if profiling shows it to be useful.